python/ray/train/mosaic/mosaic_trainer.py

```diff
@@ -0,0 +1,175 @@
+            passing datasets to the trainer and using the dataset shards in the trainer
+            loop. Instead, configure and load the datasets inside
+            ``trainer_init_per_worker`` function
+        trainer_init_config: Configurations to pass into ``trainer_init_per_worker`` as
+            kwargs. To define configurations for the composer trainer's ``fit`` call,
+            use ``"fit_config"`` key.
+        torch_config: Configuration for setting up the PyTorch backend. If set to
+            None, use the default configuration. This replaces the ``backend_config``
+            arg of ``DataParallelTrainer``. Same as in ``TorchTrainer``.
+        scaling_config: Configuration for how to scale data parallel training.
+        dataset_config: Configuration for dataset ingest.
+        run_config: Configuration for the execution of the training run.
+        preprocessor: A ray.data.Preprocessor to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A MosiacCheckpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        trainer_init_per_worker: Callable[[Optional[Dict]], Trainer],
+        *,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        trainer_init_config: Optional[Dict] = None,
+        torch_config: Optional[TorchConfig] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        dataset_config: Optional[Dict[str, DatasetConfig]] = None,
+        run_config: Optional[RunConfig] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+
+        self._validate_trainer_init_per_worker(
+            trainer_init_per_worker, "trainer_init_per_worker"
+        )
+
+        self._validate_datasets(datasets)
+
+        trainer_init_config = trainer_init_config.copy() if trainer_init_config else {}
+        if "_trainer_init_per_worker" in trainer_init_config:
+            raise ValueError(
+                "'_trainer_init_per_worker' is a reserved key in `trainer_init_config`."
+            )
+        trainer_init_config["_trainer_init_per_worker"] = trainer_init_per_worker
+
+        super().__init__(
+            train_loop_per_worker=_mosaic_train_loop_per_worker,
+            train_loop_config=trainer_init_config,
+            torch_config=torch_config,
+            scaling_config=scaling_config,
+            dataset_config=dataset_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
+
+    def _validate_trainer_init_per_worker(
+        self, trainer_init_per_worker: Callable, fn_name: str
+    ) -> None:
+        num_params = len(inspect.signature(trainer_init_per_worker).parameters)
+        if num_params != 1:
+            raise ValueError(
+                f"{fn_name} should take in at most 1 argument (`**config`), "
```

Comment:
hmm isn't `**config` technically arbitrary num of arguments ? or you mean single arg `config` as a dictionary ?

Reply:
https://github.com/ray-project/ray/pull/29237#discussion_r995119764 ?

Reply:
Should I make this change?

Reply:
yep that looks indeed simpler

Reply:
updated!